source/tutorials/learning-journey/shell-dot-nix.md

```diff
@@ -0,0 +1,163 @@
+For example, you could have a database that depends on an environment variable to set the default authentication credentials during development.
+
+Setting an environment variable in via `shell.nix` is trivial.
+Any attribute in the `mkShell` function call that `mkShell` doesn't recognize as a reserved attribute name will be set to an environment variable in the shell environment.
+The attributes that are reserved are listed in the [Nixpkgs manual][mkshell_attrs] and include `packages`, `name`, and several others.
+
+[mkshell_attrs]: https://nixos.org/manual/nixpkgs/stable/#sec-pkgs-mkShell-attributes
+
+Let's say you wanted to set the database user (`DB_USER`) and password (`DB_PASSWORD`) via environment variables in your `shell.nix` file.
+This is how that would look:
+```nix
+let
+  pkgs = import <nixpkgs> {};
+in
+  pkgs.mkShell {
+    packages = [
+      pkgs.python310
+      pkgs.curl
+    ];
+
+    env = {
+      # Database credentials
+      DB_USER = "db_user";
+      DB_PASSWORD = "super secret don't look";
+    };
+  }
+```
+
+
+Some variables are protected.
+For example, the shell prompt format for most shells is set by the `PS1` environment variable, but `nix-shell` already overrides this by default, and does not allow us to alter the `PS1` attribute directly.
+In cases where `nix-shell` prevents you from settings these environment variables directly, we can still set a new value for that environment variable by manually exporting it in the [`shellHook` attribute](https://nixos.org/manual/nixpkgs/stable/#sec-pkgs-mkShell-attributes) passed to `mkShell`.
+
+To set the shell prompt to the format `<username>@<hostname> >>> `, the `shell.nix` file would look like this:
+
+```nix
+let
+  pkgs = import <nixpkgs> {};
+in
+  pkgs.mkShell {
+    packages = [
+      pkgs.python310
+      pkgs.curl
+    ];
+
+    env = {
+      # Database credentials
+      DB_USER = "db_user";
+      DB_PASSWORD = "super secret don't look";
+    };
+
+    # Set the shell prompt format
+    shellHook = ''
+      export PS1="\u@\h >>> "
+    '';
+  }
+```
+
+
+## Startup commands
+You may want to perform some initialization before entering the shell environment (for example, maybe you want to ensure that a file exists).
+Commands you'd like to run before entering the shell environment can be placed in the `shellHook` attribute of the attribute set provided to the `mkShell` function.
+To ensure that a file `should_exist.txt` exists, the `shell.nix` file would look like this:
+
+```nix
+let
+  pkgs = import <nixpkgs> {};
+in
+  pkgs.mkShell {
+    packages = [
+      pkgs.python310
+      pkgs.curl
+    ];
+
+    env = {
+      # Database credentials
+      DB_USER = "db_user";
+      DB_PASSWORD = "super secret don't look";
+    };
+
+    # Set shell prompt format, ensure that 'should_exist.txt' exists
+    shellHook = ''
+      export PS1="\u@\h >>> "
+      touch should_exist.txt
+    '';
+  }
+```
```

Comment:
Discussed online: Provide other examples such as initializing a data directory for a database, getting secrets, etc.